newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/InstrumentationImpl.java

```diff
@@ -540,8 +572,79 @@ public void instrument(String className, String metricPrefix) {
         ServiceFactory.getClassTransformerService().addTraceMatcher(matcher, metricPrefix);
     }
 
+    @Override
+    public void instrument() {
+        final boolean check = autoInstrumentCheck.get();
+        if (check) {
+            final StackTraceElement stackTraceElement = getApplicationStackTraceElement(
+                    new Exception().fillInStackTrace().getStackTrace());
+            if (stackTraceElement != null) {
+                instrument(stackTraceElement, TraceDetailsBuilder.newBuilder().setAsync(true).build());
+            }
+        }
+        NewRelic.recordMetric("Supportability/InstrumentationImpl/instrument", check ? 1f : 0f);
```

Comment:
It looks like this might over report this metric in some cases where ` instrument(StackTraceElement stackTraceElement, TraceDetails traceDetails)` is called and would return immediately because instrumentedStackTraceElements.contains(stackTraceElement) is true.  Should we also check a returned boolean from instrument(...)?

I'm thinking this line is what we'd miss in the check.
    private void instrument(StackTraceElement stackTraceElement, TraceDetails traceDetails) {
        if (instrumentedStackTraceElements.contains(stackTraceElement)) {
            return;
        }

Would doing something like this make sense?

    private boolean instrument(StackTraceElement stackTraceElement, TraceDetails traceDetails) {
        if (instrumentedStackTraceElements.contains(stackTraceElement)) {
            return false;
        }
        ...
        ...
        ...
 And then checking again inside the first predicate `check`       
Or do I misunderstand the intent of the metric?
@meiao

Reply:
I'm performing a little trick with that metric, which is a technique we often use in custom instrumentation.  The count of the metric is how many times we hit the line of code and the value is the count of how many times `check` is true.  Summing the value give you the overall count of `check == true`, but the average of the metric tells you the percentage of time check was true.

Reply:
We got that. But we ran a test that should instrument only one method. And when we checked the metric, the sum was 11.

What happened was:
- a call was made to a method that needs to be instrumented
- that method is scheduled to be instrumented
- the time limit to instrument another method elapses
- another call to the same method is made
- [1]
- method is instrumented

in [1] the rateLimiter will say it is fine to instrument that method. Which will make the metric be sent as 1.0f. But when the instrument method is invoked, it will check that that method is in the `instrumentedStackTraceElements` and will return not scheduling another instrumentation.

Reply:
Sure.  That can happen.  But it won't actually try to reinstrument because of this, right?

         if (instrumentedStackTraceElements.contains(stackTraceElement)) {
             return;
         }
         instrumentedStackTraceElements.add(stackTraceElement);

Reply:
Correct, it wont instrument again. But the metric will be sent with 1 in that case.

Reply:
It is a little nitpicking, but the metric currently means how many times the rate limiter allowed execution thru, not how many methods were instrumented. But its name is instrumented.